contracts/rdt/RevenueDistributionToken.sol

```diff
@@ -0,0 +1,414 @@
+        _mint(shares_ = asset_, asset_, receiver_, msg.sender);
+    }
+
+    function mint(uint256 shares_, address receiver_) external virtual override nonReentrant returns (uint256 assets_) {
+        _mint(shares_, assets_ = shares_, receiver_, msg.sender);
+    }
+
+    function mintWithPermit(
+        uint256 shares_,
+        address receiver_,
+        uint256 maxAssets_,
+        uint256 deadline_,
+        uint8 v_,
+        bytes32 r_,
+        bytes32 s_
+    ) external virtual override nonReentrant returns (uint256 assets_) {
+        require((assets_ = previewMint(shares_)) <= maxAssets_, "RDT:MWP:INSUFFICIENT_PERMIT");
+
+        ERC20(asset).permit(msg.sender, address(this), maxAssets_, deadline_, v_, r_, s_);
+        _mint(shares_, assets_, receiver_, msg.sender);
+    }
+
+    function withdraw(
+        uint256 asset_,
+        address receiver_,
+        address owner_
+    ) external virtual override nonReentrant returns (uint256 shares_) {
+        _claim(receiver_);
+        _burn(shares_ = asset_, asset_, receiver_, owner_, msg.sender);
+    }
+
+    function redeem(
+        uint256 shares_,
+        address receiver_,
+        address owner_
+    ) external override returns (uint256 assets_) {
+        _claim(receiver_);
+        _burn(shares_, assets_ = shares_, receiver_, owner_, msg.sender);
+    }
+
+    function claim(address receiver_) external virtual nonReentrant returns (uint256 claimableRevenueAssets_) {
+        claimableRevenueAssets_ = _claim(receiver_);
+    }
+
+    function migrate(
+        address receiver_,
+        address xTRI_,
+        uint256 xTRIAmount_
+    ) external virtual nonReentrant returns (uint256 shares_) {
+        require(ERC20(xTRI_).transferFrom(receiver_, address(this), xTRIAmount_), "RDT:M:INSUFFICIENT_PERMIT");
+        uint256 triBalanceBefore_ = ERC20(asset).balanceOf(address(this));
+        ITriBar(xTRI_).leave(xTRIAmount_);
+        uint256 triBalanceUnstaked_ = ERC20(asset).balanceOf(address(this)) - triBalanceBefore_;
+
+        _mint(shares_ = triBalanceUnstaked_, triBalanceUnstaked_, receiver_, address(this));
+
+        emit Migrated(receiver_, xTRI_, triBalanceUnstaked_, xTRIAmount_);
+    }
+
+    /**************************/
+    /*** Internal Functions ***/
+    /**************************/
+
+    function _mint(
+        uint256 shares_,
+        uint256 triAmount_,
+        address receiver_,
+        address caller_
+    ) internal {
+        require(receiver_ != address(0), "RDT:M:ZERO_RECEIVER");
+        require(shares_ != uint256(0), "RDT:M:ZERO_SHARES");
+        require(triAmount_ != uint256(0), "RDT:M:ZERO_ASSETS");
+
+        _mint(receiver_, shares_);
+
+        uint256 freeAssetsCache = freeAssets = totalClaimableRevenueAssets();
+
+        uint256 issuanceRate_ = _updateIssuanceParams();
+
+        emit Deposit(caller_, receiver_, triAmount_, shares_);
+        emit IssuanceParamsUpdated(freeAssetsCache, issuanceRate_);
```

Comment:
why did we initialize `freeAssetsCache` can we not just use `freeAssets` here?

Reply:
ok understood, `_updateIssuanceParams` is also updating global variable `freeAssets`